Next Generation Sequencing Workflows
====================================

Source code, data, documentation and reference materials

References
----------
* "A Quick Guide to Organizing Computational Biology Projects" http://dx.doi.org/10.1371/journal.pcbi.1000424
* "A quick guide for developing effective bioinformatics programming skills." http://dx.doi.org/10.1371/journal.pcbi.1000589
* "Managing Projects with GNU Make." http://oreilly.com/catalog/make3/book/index.csp
* Binary sequence trace formats. http://www.ncbi.nlm.nih.gov/Traces/trace.cgi?cmd=show&f=formats&m=doc&s=format#ztr
* "Advanced Bash-Scripting Guide" http://tldp.org/LDP/abs/html/
* NGS glossary https://docs.google.com/spreadsheet/ccc?key=0Av8UW3JvZsgcdE9wZW1sYzlCQWFwNjBXLWMtQzZLN3c#gid=0
* NGS platforms https://docs.google.com/document/d/1rYbBPELjjezRVjkQfkulJI2jNxL5LsRuNXVv_CxCpd4/edit

Blog posts
----------
* "A Closer Look At The First PacBio Sequence Dataset" http://oelemento.wordpress.com/2011/01/03/a-closer-look-at-the-first-pacbio-sequence-dataset/
* "First look at Ion Torrent data: De novo assembly" http://pathogenomics.bham.ac.uk/blog/2011/05/first-look-at-ion-torrent-data-de-novo-assembly/
* Sequence file formats http://bioinf.comav.upv.es/courses/sequence_analysis/sequence_file_formats.html
* "Introduction to the processing of short read next generation sequencing data" http://darrenjw.wordpress.com/2010/11/28/introduction-to-the-processing-of-short-read-next-generation-sequencing-data/
* "A quick introduction to the Bioconductor ShortRead package for the analysis of NGS data" http://darrenjw.wordpress.com/2010/11/29/a-quick-introduction-to-the-bioconductor-shortread-package-for-the-analysis-of-ngs-data/
* "2012 NGS Field Guide Ð Overview" http://www.molecularecologist.com/next-gen-fieldguide/
* "Maq, BWA, and Bowtie Compared" http://massgenomics.org/2009/07/maq-bwa-and-bowtie-compared.html
* Short Read Aligner comparison http://massgenomics.org/short-read-aligners

Executables
--------
* samtools http://samtools.sourceforge.net/
* BWA http://bio-bwa.sourceforge.net/
* sff2fastq https://github.com/indraniel/sff2fastq
* fastqc http://www.bioinformatics.babraham.ac.uk/projects/fastqc/
* picard http://picard.sourceforge.net/
* bowtie http://bowtie-bio.sourceforge.net/
* maq http://maq.sourceforge.net/
* velvet http://www.ebi.ac.uk/~zerbino/velvet/
* prinseq http://prinseq.sourceforge.net/
* bfast http://bfast.sourceforge.net/
* mira http://mira-assembler.sourceforge.net/

Toolkits
--------
* BioRuby bio-ngs https://github.com/helios/bioruby-ngs/
* BioPerl https://github.com/bioperl/bioperl-live/
* BioConductor http://www.bioconductor.org/
* BioPython https://github.com/biopython/biopython/
* FAST-X toolkit http://hannonlab.cshl.edu/fastx_toolkit/
* GATK http://www.broadinstitute.org/gatk/
* BioLib https://github.com/biolib/biolib
* SOAP http://soap.genomics.org.cn/index.html

Integrated applications
-----------------------
* CLC workbench http://www.clcbio.com/
* Geneious http://www.geneious.com/
* mothur http://www.mothur.org/wiki/Main_Page

Appliances
----------
* BioLinux http://nebc.nerc.ac.uk/tools/bio-linux
* QIIME http://qiime.org/index.html

Helper tools
------------
* git http://git-scm.com/
* perl http://www.perl.org/
* python http://python.org/
* ruby http://www.ruby-lang.org/en/
* r http://www.r-project.org/

Visual workflow tools
---------------------
* Galaxy http://galaxy.psu.edu/
* Taverna http://www.taverna.org.uk/

Use cases
---------
* Taxonomic placement of metagenomic data (orchid mycorrhiza)
* Alignment against reference genome (plasmodium) 
* Comparative analysis of multicellular genomic data
