Next Generation Sequencing Workflows
====================================

Source code, data, documentation and reference materials

References
----------
* "A Quick Guide to Organizing Computational Biology Projects" http://dx.doi.org/10.1371/journal.pcbi.1000424
* "A quick guide for developing effective bioinformatics programming skills." http://dx.doi.org/10.1371/journal.pcbi.1000589
* "Managing Projects with GNU Make." http://oreilly.com/catalog/make3/book/index.csp

Blog posts
----------
* "A Closer Look At The First PacBio Sequence Dataset" http://oelemento.wordpress.com/2011/01/03/a-closer-look-at-the-first-pacbio-sequence-dataset/
* "First look at Ion Torrent data: De novo assembly" http://pathogenomics.bham.ac.uk/blog/2011/05/first-look-at-ion-torrent-data-de-novo-assembly/
* Sequence file formats http://bioinf.comav.upv.es/courses/sequence_analysis/sequence_file_formats.html
* "Introduction to the processing of short read next generation sequencing data" http://darrenjw.wordpress.com/2010/11/28/introduction-to-the-processing-of-short-read-next-generation-sequencing-data/
* "A quick introduction to the Bioconductor ShortRead package for the analysis of NGS data" http://darrenjw.wordpress.com/2010/11/29/a-quick-introduction-to-the-bioconductor-shortread-package-for-the-analysis-of-ngs-data/

Binaries
--------
* SAM/BAM source code http://samtools.sourceforge.net/
* BWA source code http://bio-bwa.sourceforge.net/

Toolkits
--------
* BioRuby bio-ngs https://github.com/helios/bioruby-ngs/
* BioPerl https://github.com/bioperl/bioperl-live/
* BioConductor http://www.bioconductor.org/
* BioPython https://github.com/biopython/biopython/

Use cases
---------
* Phylogenetic placement of metagenomic data (orchid mycorrhiza)
* Typing of pathogens
* Comparative analysis of multicellular genomic data
